src/Symfony/Component/DependencyInjection/Compiler/DecoratorServicePass.php

```diff
@@ -62,6 +66,40 @@ public function process(ContainerBuilder $container)
             }
 
             $container->setAlias($inner, $id)->setPublic($public)->setPrivate($private);
+            $this->autowire($container, $definition, $renamedId);
+        }
+    }
+
+    private function autowire(ContainerBuilder $container, Definition $definition, string $renamedId): void
+    {
+        if (
+            !$definition->isAutowired() ||
+            null === ($innerClass = $container->findDefinition($renamedId)->getClass()) ||
+            !($reflectionClass = $container->getReflectionClass($definition->getClass())) ||
+            !$reflectionMethod = $reflectionClass->getConstructor()
+        ) {
+            return;
+        }
+
+        $innerIndex = null;
+        foreach ($reflectionMethod->getParameters() as $index => $parameter) {
+            if (
+                null === ($type = ProxyHelper::getTypeHint($reflectionMethod, $parameter, true)) ||
+                !is_a($innerClass, $type, true)
+            ) {
+                continue;
+            }
+
+            if (null !== $innerIndex) {
+                // There is more than one argument of the type of the decorated class
+                return;
```

Comment:
Maybe this case should result in an exception with helpful feedback for the developer. The developer obviously tried to autowire the decorator, so I think we should tell them why it fails.

Reply:
It would be a BC break: it currently try to autowire these services "the regular way", if we throw an exception, it would not be the case anymore.